### PR TITLE
feat(logs): add LogRecord.SeverityText accessor

### PR DESCRIPTION
**Pull request by jpkrohling**
## Motivation

`LogRecord.SeverityNumber()` exists; `severity_text` (field 3) did not. sage hand-rolls the walk today in `internal/event/wire.go` as `logRecordSeverityText(data []byte) (string, error)`. This adds the symmetrical accessor.

Closes E-2944, sub-issue 4 of E-2940.

## Approach

```go
func (r LogRecord) SeverityText() ([]byte, error)
```

The design decision worth reviewing: **`SeverityText` reads out of the existing schema-aware `parseLogRecordSeverity` walk rather than calling `extractLastBytesField`.**

A second, shallower parser over the same message is how two accessors come to disagree about whether the same bytes are valid. A field-3-only walk (what a consumer hand-rolls, and what `extractLastBytesField` would give) accepts a malformed body or attribute that `SeverityNumber` rejects — so a consumer reading both fields sees one succeed and one fail on identical input. Sharing the walk makes that impossible by construction; `TestLogRecordSeverity_MalformedParity` pins it.

Contract:

- returns `nil` when absent, a non-nil zero-length slice when present-but-empty — a distinction pdata cannot represent;
- repeated occurrences resolve last-value-wins, as protobuf/pdata do for a singular scalar;
- returned slice aliases the request buffer with capacity clamped, so a caller's `append` cannot overwrite adjacent record fields.

`AGENTS.md` previously told contributors to reach for `extractLastBytesField` for *any* singular scalar — literally the shallow second parser this change exists to avoid. It now records the third branch.

## Compatibility

Additive. No exported signature changed, no behavior changed for existing accessors. Every consumer upgrades with a plain dependency bump.

## Validation

```bash
go build ./...
go vet ./...
go test -v -race ./...
go test -race -shuffle=on -count=1 ./...
git diff --check
```

All pass. Per `CONTRIBUTING.md`, `go mod tidy -diff` is not gated on (pre-existing `go.sum` checksum history); no dependency changed here.

New coverage in `log_iteration_test.go`:

- pdata differentials: absent, present-but-empty, populated, repeated (last wins), repeated resolving *to* an empty last occurrence and *from* an empty first one, unknown scalar fields skipped;
- `TestLogRecordSeverity_MalformedParity` — 9 malformed records, each asserted to fail `SeverityText`, `SeverityNumber` **and** pdata. This replaces `TestLogRecordSeverityNumber_PdataMalformedParity`, whose 6 fixtures it contains verbatim plus 3 new rows, and additionally asserts the second accessor;
- complete-message semantics: trailing corruption and a trailing wrong-wire-type occurrence still error after a valid early value;
- view/allocation contract: capacity clamp, aliasing proven by mutating the buffer, and `testing.AllocsPerRun` == 0;
- `TestLogRecordSeverity_PdataDivergence` — the five inputs where the shared walk and pdata disagree, asserted in both directions, with the two accessors still required to agree with each other.

## Benchmarks

Paired, interleaved single-arm invocations of prebuilt binaries (not `go test -count=N`, which does not interleave arms). i7-11800H, linux/amd64, Go 1.25.12.

**Cost of the shared walk on the existing `SeverityNumber` hot path** — fixture emits no `severity_text`, so this measures only the added branching and return:

| Revision | ns/op (median of 15) | B/op | allocs/op |
|---|---:|---:|---:|
| merge base | 78,744 | 488 | 15 |
| this change | 80,749 | 488 | 15 |

**Median paired delta +2.67%, slower in 14 of 15 rounds, allocations unchanged.** A second independent 15-round session measured +2.00% (12/15). Read the cost as **+2 to +3%**: single rounds range −8.7% to +5.8% under desktop load, so sign consistency is the evidence and the magnitude is stable only to about a percentage point.

**Against sage's hand-rolled walk** (copied verbatim as the comparison arm):

| Benchmark | ns/op (median of 9) | B/op | allocs/op |
|---|---:|---:|---:|
| `_HandRolled` | 40,215 | 2,368 | 514 |
| `SeverityText` | 69,185 | 368 | 14 |

~72% slower, 500 allocations cheaper per batch. The arms are deliberately **not** equivalent work: sage's stops at field 3 and materializes a `string`; this validates every known LogRecord field and returns a view. Both directions are documented in `docs/BENCHMARKS.md` — a consumer that wants the cheap shallow read and does not need `SeverityNumber` parity is better served by its own walk, and should say so on the migration issue.

## Risks

- **Accessor CPU cost, disclosed above.** +2 to +3% on the severity-classification path; allocations unchanged. Not a hot-path regression in allocation terms.
- **Two walks for both fields.** Each accessor runs the walk once, so reading number + text costs two. No combined accessor added — the walk already produces both values so exposing the pair needs no new machinery, but per E-2940's flattened-convenience rule that needs a measured hot path. Concrete trigger recorded: sage adopting `SeverityText` and reading both fields per record.
- **API compatibility:** none, additive.
- **Rollout:** none required. Library-only change; no release or consumer migration in this PR. sage's migration is tracked separately.

## Reviewer notes

Two things I deliberately did **not** do, with the measurements behind them:

1. **Split `case 3, 12` into separate arms** for field-specific error messages. Measured ~0.8pp *slower* split; the shared arm is also house style (`case 9, 10` discriminates the same way). The current error text does not name `severity_text` specifically — flagging in case you'd trade the 0.8pp for the diagnostic.
2. **Reshape `parseLogRecordSeverity` to return a `logRecordFields` struct** instead of threading a third return value. Fair argument that it is the same churn today and zero churn for the next LogRecord accessor. Skipped because this function's cost is measurably sensitive to loop-carried state layout, so reshaping means re-running the full paired benchmark for a benefit that only lands on a hypothetical third accessor. Happy to do it as a follow-up if you'd rather have the shape now.

## Known divergences recorded, not introduced

Adversarial review turned up three classes of wire-vs-pdata disagreement, all pre-existing (`SeverityNumber` behaves identically on `main`) and all reachable only on malformed or adversarial input:

| Input | Wire path | pdata |
|---|---|---|
| Well-formed group in an unknown LogRecord field, or inside the body's `AnyValue` | accepts | rejects |
| Varint overflowing `uint64` (10 bytes, final byte ≥ 2), including a `severity_text` length prefix | rejects | accepts (truncated) |
| Field number above `MaxInt32` truncating positive | rejects | accepts |

The last two are `protowire` being stricter than pdata's generated unmarshal, which silently truncates both. `TestLogRecordSeverity_PdataDivergence` pins all five so a change in either direction fails a test rather than silently altering consumer behavior, and `operations.md` documents them for consumers that pair the wire path with a pdata fallback.

## Adversarial review

Per E-2940's gate. It could not break the accessor:

- **Parity** (the change's central claim): 6.9M structured-grammar and 2.0M raw-byte fuzz executions asserting `(numErr==nil) == (txtErr==nil)` — zero failures.
- **No regression in `SeverityNumber`**: the pre-change parser was copied in as a second implementation and differentially fuzzed over ~11M inputs comparing both error-ness and value — zero disagreements.
- **Aliasing/capacity**: appending 32 bytes through the returned view left the enclosing request bytes byte-identical across populated, empty, repeated and multi-record buffers.
- **Zero-alloc** holds on rich records (body + 3 attributes + trace/span IDs + flags + event_name), not just the bare fixture.

It did find five real defects, all in docs and coverage, fixed in the second commit: a dangling test reference, the incomplete divergence note above, a benchmark recipe whose "before" arm was a `git stash` no-op that would have measured 0%, the understated regression figure, and a missing test for a repeated `severity_text` whose last occurrence is empty — the case where a `len(value) > 0` guard would silently return the earlier value.

## Agent involvement

Per `CONTRIBUTING.md`: this change was implemented with Claude Code. The implementation, tests, docs and all benchmark runs were agent-produced; findings from four cleanup sub-agents and one adversarial sub-agent were verified against the code before being applied or dismissed. A human must review and take ownership before merge.

## Review status

CodeRabbit raised 4 actionable comments across two passes; all are fixed and all four threads are resolved.

- `example_test.go`: the example returned early on an accessor error without invoking `scopesErr`/`resourcesErr`. No error was swallowed, but examples are executable documentation and this one taught a pattern that violates the repository-wide iterator contract. Now one function per level, each checking its own closure.
- `operations.md`: the divergences were documented in prose only — now pinned by a test.
- `README.md`: restored the last-value-wins clause I had trimmed as duplication; it belongs in the API summary.
- `log_iteration_test.go`: I had called the rejected varints "non-canonical". They are *overflowing* — a 10-byte varint carries at most bit 63 in its final byte, so a final byte ≥ 2 exceeds `uint64`. Renamed and corrected in the docs too.

### CODE REVIEW

File: example_test.go [hunk 1]
```diff
@@ -268,6 +263,64 @@ func ExampleResourceLogs_Resource() {
...
+				if err != nil {
+					return err
+				}
+				number, err := record.SeverityNumber()
+				if err != nil {
+					return err
+				}
+				text, err := record.SeverityText()
+				if err != nil {
+					return err
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply example_test.go]: Valid, fixed in 4f33b4c. The traversal is now one function per level (`printLogSeverities` / `printResourceLogSeverities` / `printScopeLogSeverities`), each checking its own error closure before returning the preserved failure. `printScopeLogSeverities` needs none, since `LogRecordsSeq` yields its errors inline rather than deferring them — noted in a comment so the asymmetry does not read as an omission.

Worth fixing even though no error was actually swallowed (the accessor error was still returned): examples are executable documentation here, and this one was teaching a pattern that violates the repository-wide contract.

File: operations.md [hunk 2]
```diff
@@ -45,6 +46,40 @@ release time, per "Release and production analysis" below. Confirm the canary
...
+| --- | --- | --- |
+| Well-formed group in an unknown LogRecord field, or inside the body's `AnyValue` | accepts | rejects (`unexpected EOF`) |
+| Non-canonical 10-byte varint (10th byte ≥ 2), in any varint field including a `severity_text` length prefix | rejects | accepts |
+| Field number above `MaxInt32` whose `int32` truncation is positive | rejects (`malformed protobuf tag`) | accepts (truncates, then skips) |
+
+The first row is the mirror of the unclosed-group row above. The last two come
+from `protowire` being stricter than pdata's generated unmarshal: `protowire`
+enforces canonical varint encoding and rejects out-of-range field numbers,
+where the gogo-generated loop does neither. A consumer cannot conclude "pdata
+would also reject this" from a wire-path error, nor the reverse.
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply operations.md]: Agreed, added `TestLogRecordSeverity_PdataDivergence` in 4f33b4c. It covers all five documented cases with the exact reproducers: unknown-field group, group inside the body `AnyValue`, non-canonical 10-byte varint in `severity_number` and as the `severity_text` length prefix, and a field number above `MaxInt32` that truncates positive.

It asserts the documented outcome in both directions rather than just "a parser errors" — `wireRejects` selects whether the wire path must error and pdata must accept, or the reverse. It also asserts that `SeverityText` and `SeverityNumber` still agree with each other on every one of these inputs, since the divergence is wire-versus-pdata and accessor parity must survive it.

`operations.md` no longer says "no test covers them".

File: README.md [hunk 3]
```diff
@@ -256,13 +257,20 @@ func (s ScopeLogs) LogRecordsSeq(yield func(LogRecord, error) bool) // zero-allo
...
 func (r Resource) Attributes() (iter.Seq[KeyValue], func() error)
 func (r Resource) AttributesSeq(yield func(KeyValue, error) bool)
 func (r Resource) StringAttribute(key string) ([]byte, bool, error)
 ```
 
+`LogRecord.SeverityText` returns raw bytes aliasing the request buffer, and
+`nil` when `severity_text` is absent, which distinguishes an absent field from
+a present empty one. It shares `SeverityNumber`'s schema-aware walk of the
+whole LogRecord, so reading both fields costs two walks; see
+[docs/DESIGN.md](docs/DESIGN.md) for why they share one parser.
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply README.md:273]: Valid, restored in 4f33b4c. I had trimmed that clause as duplication with `docs/specification.md`, which was the wrong call — the README block is the API summary, and last-value-wins for a repeated singular scalar is a contract detail a reader of the summary needs rather than a cross-reference they should have to follow.

File: log_iteration_test.go [hunk 4]
```diff
@@ -71,7 +73,183 @@ func TestLogRecordSeverityNumber_CompleteMessageSemantics(t *testing.T) {
...
+			// 10-byte varint whose final byte is >= 2: protowire requires
+			// canonical encoding, pdata's generated loop does not.
+			name:        "non-canonical varint in severity_number",
+			record:      LogRecord{0x10, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x02},
+			wireRejects: true,
+		},
+		{
+			name:        "non-canonical varint as the severity_text length",
+			record:      LogRecord{0x1a, 0x81, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x02, 'X'},
+			wireRejects: true,
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply log_iteration_test.go]: Correct, fixed in c153172. A 10-byte varint carries at most bit 63 in its final byte, so a final byte >= 2 encodes a value `uint64` cannot hold — `protowire` rejects the overflow, while pdata's generated loop stops shifting at 64 bits and keeps the truncated value. "Non-canonical" would mean a representable value spelled with redundant bytes, which these fixtures do not exercise.

Renamed both cases to "overflowing varint ...", corrected the explanatory comment, and updated the matching row and prose in `operations.md` (which now also says pdata accepts it *truncated*, rather than implying it round-trips).

**Review by jpkrohling — COMMENTED** [review]
human review ok, lgtm

we do need to clean up those docs in a follow up pr once we finish this set of tasks
